src/com/amazon/ion/impl/IonReaderContinuableApplicationBinary.java

```diff
@@ -239,6 +514,43 @@ String getSymbol(int sid) {
         return symbols[localSid];
     }
 
+    /**
+     * Creates a SymbolToken representation of the given symbol ID.
+     * @param sid a symbol ID.
+     * @return a SymbolToken.
+     */
+    private SymbolToken getSymbolToken(int sid) {
+        int symbolTableSize = localSymbolMaxId + firstLocalSymbolId + 1; // +1 because the max ID is 0-indexed.
+        if (symbolTokensById == null) {
+            symbolTokensById = new ArrayList<>(symbolTableSize);
+        }
+        if (symbolTokensById.size() < symbolTableSize) {
+            for (int i = symbolTokensById.size(); i < symbolTableSize; i++) {
+                symbolTokensById.add(null);
+            }
+        }
+        if (sid >= symbolTableSize) {
+            throw new IonException("Symbol ID exceeds the max ID of the symbol table.");
+        }
+        SymbolToken token = symbolTokensById.get(sid);
+        if (token == null) {
+            String text = getSymbolString(sid, imports, symbols);
+            ImportLocation importLocation = null;
+            if (text == null) {
+                // Note: this will never be a system symbol.
+                if (sid > 0 && sid < firstLocalSymbolId) {
+                    importLocation = imports.getImportLocation(sid);
+                } else {
+                    // All symbols with unknown text in the local symbol range are equivalent to symbol zero.
+                    sid = 0;
+                }
+            }
+            token = new SymbolTokenWithImportLocation(text, sid, importLocation);
+            symbolTokensById.set(sid, token);
+        }
+        return token;
```

Comment:
So you have either `text` or an import location, but not both? I'm confused by this. Is this because the import location is only relevant if the imported symbol table didn't fill its declared range?

EDIT: I see this is clarified by later comment and by #553 .

Suggestion removes a level of nesting.

```suggestion
        SymbolToken token = symbolTokensById.get(sid);
        if (token != null) {
            return token;
        }

        String text = getSymbolString(sid, imports, symbols);
        ImportLocation importLocation = null;
        if (text == null) {
            // Note: this will never be a system symbol.
            if (sid > 0 && sid < firstLocalSymbolId) {
                importLocation = imports.getImportLocation(sid);
            } else {
                // All symbols with unknown text in the local symbol range are equivalent to symbol zero.
                sid = 0;
            }
        }
        token = new SymbolTokenWithImportLocation(text, sid, importLocation);
        symbolTokensById.set(sid, token);
        return token;
```

Reply:
I'm happy to move the de-nesting suggestion to #553 , adjusted for absence of SymbolTokenWithImportLocation.